Add BenchmarkDotNet Benchmarks

Turnerj opened the pull request:
This starts addressing #2 by adding BenchmarkDotNet to a new project with various LINQ benchmarks.

Currently:
- Sum (faster)
- Where (slower)
- Skip (slower)
- Take (slower)
- ToList (same)
- ToDictionary (6% - 10% slower)
- OrderBy (6% - 10% slower, twice the allocations - note: this test, the data is already ascending)
- OrderByDescending (15% - 18% slower, twice the allocations - note: this test, the data is ascending)
- Min (~55% faster)
- Max (~85% faster - even faster than a ForLoop)
- Average (~40% faster)
- Select (0% - 17% faster)
- Concat (45% - ~7000% slower - falls down quite far for large number of concats)

Turnerj committed:
Added BenchmarkDotNet with sum benchmark

Turnerj commented:
Here are the results from that first benchmark on my machine. About 70% faster than `System.Linq` but 3x the allocations.

```
BenchmarkDotNet=v0.11.5, OS=Windows 10.0.18362
Intel Core i7-6700HQ CPU 2.60GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.0.100-preview7-012821
  [Host] : .NET Core 3.0.0-preview7-27912-14 (CoreCLR 4.700.19.32702, CoreFX 4.700.19.36209), 64bit RyuJIT  [AttachedDebugger]
  Core   : .NET Core 3.0.0-preview7-27912-14 (CoreCLR 4.700.19.32702, CoreFX 4.700.19.36209), 64bit RyuJIT

Job=Core  Runtime=Core
```

|         Method | NumberOfItems |       Mean |     Error |    StdDev | Ratio |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------- |-------------- |-----------:|----------:|----------:|------:|-------:|------:|------:|----------:|
|        ForLoop |          1000 |   1.304 us | 0.0148 us | 0.0139 us |  0.18 |      - |     - |     - |         - |
|  SystemLinqSum |          1000 |   7.394 us | 0.0972 us | 0.0909 us |  1.00 | 0.0153 |     - |     - |      48 B |
| CisternLinqSum |          1000 |   2.118 us | 0.0347 us | 0.0324 us |  0.29 | 0.0458 |     - |     - |     152 B |
|                |               |            |           |           |       |        |       |       |           |
|        ForLoop |         10000 |  13.296 us | 0.0935 us | 0.0875 us |  0.18 |      - |     - |     - |         - |
|  SystemLinqSum |         10000 |  73.732 us | 0.9208 us | 0.8613 us |  1.00 |      - |     - |     - |      48 B |
| CisternLinqSum |         10000 |  20.269 us | 0.3089 us | 0.2890 us |  0.27 | 0.0305 |     - |     - |     152 B |
|                |               |            |           |           |       |        |       |       |           |
|        ForLoop |         20000 |  26.492 us | 0.3144 us | 0.2941 us |  0.18 |      - |     - |     - |         - |
|  SystemLinqSum |         20000 | 148.430 us | 2.0415 us | 1.9097 us |  1.00 |      - |     - |     - |      48 B |
| CisternLinqSum |         20000 |  40.581 us | 0.3068 us | 0.2870 us |  0.27 |      - |     - |     - |     152 B |

Turnerj committed:
Added Where benchmark

Turnerj commented:
Where Benchmarking

|         Method | NumberOfItems |         Mean |      Error |     StdDev | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------- |-------------- |-------------:|-----------:|-----------:|------:|--------:|-------:|------:|------:|----------:|
|        ForLoop |         10000 |     6.680 us |  0.0579 us |  0.0542 us |  0.25 |    0.00 |      - |     - |     - |         - |
|  SystemLinqSum |         10000 |    26.290 us |  0.4635 us |  0.4336 us |  1.00 |    0.00 | 0.0305 |     - |     - |     112 B |
| CisternLinqSum |         10000 |    35.249 us |  0.4828 us |  0.4280 us |  1.34 |    0.02 | 0.0610 |     - |     - |     216 B |
|                |               |              |            |            |       |         |        |       |       |           |
|        ForLoop |       1000000 |   722.537 us | 11.2168 us | 10.4922 us |  0.28 |    0.01 |      - |     - |     - |         - |
|  SystemLinqSum |       1000000 | 2,597.573 us | 42.3519 us | 37.5439 us |  1.00 |    0.00 |      - |     - |     - |     112 B |
| CisternLinqSum |       1000000 | 3,386.119 us | 59.8689 us | 56.0014 us |  1.30 |    0.02 |      - |     - |     - |     216 B |

Turnerj commented:
Sum Benchmarking

|         Method | NumberOfItems |         Mean |      Error |     StdDev | Ratio |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------- |-------------- |-------------:|-----------:|-----------:|------:|-------:|------:|------:|----------:|
|        ForLoop |          1000 |     1.200 us |  0.0227 us |  0.0223 us |  0.26 |      - |     - |     - |         - |
|    ForEachLoop |          1000 |     1.174 us |  0.0204 us |  0.0191 us |  0.25 |      - |     - |     - |         - |
|  SystemLinqSum |          1000 |     4.636 us |  0.0914 us |  0.0810 us |  1.00 | 0.0076 |     - |     - |      32 B |
| CisternLinqSum |          1000 |     1.254 us |  0.0249 us |  0.0255 us |  0.27 | 0.0095 |     - |     - |      32 B |
|                |               |              |            |            |       |        |       |       |           |
|        ForLoop |       1000000 | 1,256.538 us | 23.5863 us | 23.1649 us |  0.27 |      - |     - |     - |         - |
|    ForEachLoop |       1000000 | 1,249.775 us | 23.8536 us | 27.4698 us |  0.27 |      - |     - |     - |         - |
|  SystemLinqSum |       1000000 | 4,638.481 us | 79.0828 us | 73.9741 us |  1.00 |      - |     - |     - |      32 B |
| CisternLinqSum |       1000000 | 1,256.507 us | 20.9115 us | 19.5606 us |  0.27 |      - |     - |     - |      32 B |

Turnerj committed:
Dropped "where" from sum benchmark

Turnerj committed:
Tweaked method names for benchmarks

Turnerj committed:
Added ToList benchmark

Turnerj committed:
Added a numeric benchmark base

This removes some of the bootstrapping code that is shared by all the benchmarks

Turnerj committed:
Added skip benchmark

Turnerj committed:
Added take benchmark

Turnerj committed:
Added ToDictionary benchmark

Turnerj committed:
Drop "For" loop to use only "ForEach"

Turnerj committed:
Dropped constructor test and improved "ForLoop" test

Turnerj committed:
Added Max benchmark

Turnerj committed:
Added Min benchmark

Turnerj committed:
Updated exception message

Turnerj committed:
Added average benchmark

Turnerj committed:
Added orderby benchmarks

manofstick commented:
@Turnerj 

Cheers for adding these. How do you want to play this? I'm happy to merge as we you go, creating new PRs as you more (me being a bit presumptuous here!), or if you want to do more in this PR I'm happy with that too.

You are right that this creates more garbage - part and parcel with this method, I'm trying to think of possible was to limit some... I'll think, but good to get some numbers.

...some additions to performance could be some smaller collections - Cistern will be bad for these, because of some additional startup overhead - but how bad? 

Also consuming Cistern's IEnumerables's with a foreach loop rather than just using the aggregation functions. ie. like where you have:

    System.Linq.Enumerable.First(System.Linq.Enumerable.Where(Numbers, n => n == NumberOfItems));

Could be tested also as

    int first;
    foreach(var x in System.Linq.Enumerable.Where(Numbers, n => n == NumberOfItems))
    {
        first = x;
        break;
    }

Also testing where the input source is a List, an IEnumerable created by the c# compiler (i.e. a yield function), etc.

Anyway, any work is *greatly* appreciated! Thanks again.

Turnerj commented:
Hey, I'll probably just polish up the ones I've got for now as they seem to cover the major areas - happy to add any others when needed and/or expand on any when you've done any specific optimisations.

While some of the allocations may be unavoidable, it might be worth (later) looking at directly allocating memory through the `Marshal` class and de-allocating ourselves which would avoid hits against the GC. It may even help with speed in some cases too.

Pretty easy to add lower iterations cases too - just changing the params attribute values in `NumericBenchmarkBase`. I had them initially but changed it when I saw some of your other benchmarks.

I agree about some of the tests with switching out the nested aggregation functions for directly calling the enumerable - that way the test more specifically targets one type of change.

With the `List` vs `IEnumerable` vs compiler-generated `IEnumerable`, yeah that would be interesting to see. For `List`, I figure it might be because you can target specific optimisations to `IList` etc but not sure if there would be anything particularly different from the compiler-generated `IEnumerable`.

Turnerj committed:
Added lower item count tests

Turnerj committed:
Switched nested aggregation calls for foreach

Turnerj committed:
Added select benchmark

Turnerj commented:
@manofstick , I think I'm happy with the benchmarks now. I've added a lower-bound to the tests (100 items), removed the `First` call from the `WhereBenchmark` and added a select benchmark.

It might be worth adding some benchmarks later which do a series of specific aggregations in a row but for now, I'll leave that.

I was thinking of doing a few more PRs - I noticed the tests are failing (at least for me) so was going to get to that and maybe change some things like the target framework (move the library to .NET Standard 2.1, move the tests to run .NET Core 3 as well as 2.2).

I had a look at the actual LINQ code with consumables and such, while it might just be me being tired, seems a little too complex for me to dive in and help out for now. (I get the gist of it, it builds a chain of activities that execute etc, just seems to be a lot of code to do it 🤷‍♂️)

manofstick commented:
The core of it is relatively simple, but then the layers of optimisations have added to the approachability of the code base. And yes, .net standard 2.1 is the plan, but happy to leave it at core 2.2 until standard 2.1 is released properly, just for ease of tooling. There are 6 tests failing and there are some that I have set to ignore. I think I have created issues around them. None are trivial, unfortunately (I don't think). Anyway, hopefully I'll get some more time on the weekend...

Turnerj committed:
Added concat benchmark
